### PR TITLE
Hide concurrency limit reached exception log.

### PR DESCRIPTION
**Pull request by asdacap**
Addresses #4899

## Changes:
- Swallow concurrency limit reached exception.
- Slight refactor of sync dispatcher dispatch call.
- Reduces old request timeout limit.

Screenshot from 2022-11-16 21-39-23

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [X] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [ ] Yes
- [X] No

**Comments about testing , should you have some** (optional)

- Chiado sync is running.

## Further comments (optional)

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/Nethermind/Nethermind.Synchronization/ParallelSync/SyncDispatcher.cs [hunk 1]
```diff
@@ -160,6 +128,49 @@ public async Task Start(CancellationToken cancellationToken)
...
+        private async Task DoDispatch(CancellationToken cancellationToken, PeerInfo? allocatedPeer, T request,
+            SyncPeerAllocation allocation)
+        {
+            try
+            {
+                await Dispatch(allocatedPeer, request, cancellationToken);
+            }
+            catch (ConcurrencyLimitReachedException)
+            {
+                if (Logger.IsDebug) Logger.Debug($"{request} - concurrency limit reached. Peer: {allocatedPeer}");
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/ParallelSync/SyncDispatcher.cs:140]: We just ignore it, will it be re-dispatched later?

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Network/P2P/MessageDictionary.cs [hunk 2]
```diff
@@ -55,7 +56,7 @@ public void Send(Request<T66Msg, TData> request)
     {
         if (_requestCount >= MaxConcurrentRequest)
         {
-            throw new InvalidOperationException("Concurrent request limit reached");
+            throw new ConcurrencyLimitReachedException($"Concurrent request limit reached. Message type: {typeof(TMsg)}");
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/MessageDictionary.cs:59]: Btw this limit is per message? Per Peer?

**Comment by kamilchodola** [comment]
@asdacap @LukaszRozmej 
Can we have more detailed info why this log even started to appear?
I know this may be some kind of timeout - but before we do not have such ones. We can't just simply hide log and pretend like nothing bad happened.

**Comment by kamilchodola** [comment]
image
Before I have 0, later on just a few - now full page

**Review by MarekM25 — CHANGES_REQUESTED** [review]
Guys, @kamilchodola said that we didn't have these logs on version 1.14.5. Do we know why we have them now?

**Comment by asdacap** [comment]
This is part of eth66. Limit is 32 per peer. Still looking at why it can get so high.

**Comment by LukaszRozmej** [comment]
Related PR: https://github.com/NethermindEth/nethermind/pull/4758

File: src/Nethermind/Nethermind.Synchronization/ParallelSync/SyncDispatcher.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/ParallelSync/SyncDispatcher.cs:140]: We already ignore it currently... so yes?

File: src/Nethermind/Nethermind.Network/P2P/MessageDictionary.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Network/P2P/MessageDictionary.cs:59]: Per peer.

**Comment by asdacap** [comment]
I've reduced the request cleanup threshold from 60s to 10s, which reduces the concurrency limit. After recording and printing stack trace, they all came from peer refresher, so its likely that one peer is not catching up to head, and not responding to get header request, causing the request to be pending and considered concurrent.

This is likely to happen even pre eth66, we just dont limit it, and the peer would have just timeout everything.

For reference, locally I once measure timeout rate for get block bodies to be about 20%. So, this is likely not a breaking problem.

**Comment by kamilchodola** [comment]
@asdacap Any plans for new/updated unit tests?

**Comment by asdacap** [comment]
No plan

**Comment by asdacap** [comment]
That PR does not change behaviour except logging and config.

**Review by MarekM25 — APPROVED** [review]

**Review by kamilchodola — APPROVED** [review]
Wasn't able to reproduce this problem again :) Reproduced it on pure master, and on the same machine tried multiple times on this branch and no issue now.
Good job :)

**Pull request merged**